choose report width according to window-text-width

akater opened the pull request:
Register reports need as much width as possible, at least on smaller displays.

Notes:
- I replace end-of-line because command string may have multiple lines.
- Console emacs adjustment was chosen empirically.

Side notes:
- I don't use any margins or fringes in emacs. Further tweaks might be needed due to this.
- I guess I use longer account names than usual. The patch was motivated by this.

purcell commented:
There's `window-text-width`, in recent Emacsen at least, so no guesses should be necessary.

akater commented:
I tried to set `--columns` to `(window-text-width)` without further
adjustment, and the report output in tty emacs still looks ugly. (BTW,
in my case window is split horizontally, i.e. reports appear on the
right.) But yes, `window-text-width` is more appropriate.

purcell commented:
Thanks. Please don't `[ci skip]` all these changes - build feedback is important if code is to be merged.

jwiegley commented:
Agreed.

akater commented:
It would be more appropriate to add `--comment` after each `%(binary)` but the latter string is itself varying. Maybe I'll do this tomorrow, maybe in a week.

akater commented:
The mechanism in its current state will fail if provided path to ledger binary contains whitespace, which is pointed out in the source. I don't know how to implement an alternative that would work around that. Overall, an operation that modifies command line can never be completely fail-safe, and I don't find this particular issue important.

As of Travis check, I don't understand what to do about it. The first error line in the log is,
`E: Failed to fetch http://ppa.launchpad.net/boost-latest/ppa/ubuntu/dists/trusty/main/binary-amd64/Packages  404  Not Found`
I don't think I have any control over this. My patch is unrelated and trivial.

purcell commented:
> As of Travis check, I don't understand what to do about it. The first error line in the log is,
> E: Failed to fetch http://ppa.launchpad.net/boost-latest/ppa/ubuntu/dists/trusty/main/binary-amd64/Packages 404 Not Found
> I don't think I have any control over this. My patch is unrelated and trivial.

Yes, that failure in the build is our problem to address, but it's still preferable for the build to run.

Regarding the code, I wasn't quite clear why the regex replacement based on the binary path was necessary. It seems like the following (shorter) snippet might do the same thing:
```el
 (let ((data-pos (point))
        (register-report (string-match " reg\\(ister\\)? " cmd))
        files-in-report
        columns (- (window-width) (if (display-graphic-p) 0 2)))
    (shell-command
     (concat cmd " --columns " (int-to-string columns)
             (when (and register-report
                        ledger-report-links-in-register
                        ;; --subtotal does not produce identifiable transactions, so don't
                        ;; prepend location information for them
                        (not (string-match "--subtotal" cmd)))
               " --prepend-format='%(filename):%(beg_line):'"))
     t nil)
```

It also seems like `window-text-width` is very recent, so we'd need to use `window-width` instead, as per the above snippet.

Another potential issue is that the manual notes that `--columns` affects the register report only, so I'm wondering if it's safe to include that argument when the report is not a register report.

Can you try the above out please?

akater commented:
>Regarding the code, I wasn't quite clear why the regex replacement
>based on the binary path was necessary. 

There might be multiple calls to ledger in a single “command”. E.g., I am using such a (balance) report. One call with a specific `-X` option, another one without.

>
>It also seems like `window-text-width` is very recent, so we'd need to
>use `window-width` instead, as per the above snippet.

How about `fboundp` check for `window-text-width`?

>Another potential issue is that the manual notes that `--columns`
>affects the register report only, so I'm wondering if it's safe to
>include that argument when the report is not a register report.

I tried all default reports in ledger-mode, with no issues.

akater commented:
Tested with `window-width` and `window-text-width`, with fringes and with scrollbars. Results looked fine in both tty emacs and GUI emacs.

akater commented:
The same (conceptually) number is utilized above in `(make-string (- window-width 1) ?=)`. Don't know why subtract 1 and not 0 but brought the introduced local variable into accordance with the choice made before.

akater commented:
Sorry to bother. Just to make it clear, are there any unresolved problems with this PR?

Summary of issues:
#### why regex replacement
Consider the following ledger-report entry:

```
("cash-and-rides" "%(binary) -f %(ledger-file) reg cash
%(binary) -f %(ledger-file) -X rides reg travelcard")
```
Appending options to the tail of the string would apply said options to the second command only, and the output would look ugly. One can also come up with examples of useful commands which don't make use of ledger at all in their last lines.

#### window-text-width is recent
I added an `fboundp` check. Didn't test with Emacs other than 25.3. The current version already uses `window-width` for the same purpose (output formatting), and I adjusted the tweaks accordingly, so the patch conforms to what's already there.

#### --columns option for non-register report
Neither Balance report nor Cleared report choke on `--columns`. I must admit though, I don't use Cleared report at all.

If I need to improve the patch further, please tell. I'd like to treat #20 in a similar way (pre-packaged command line options, maybe activated by a `defcustom`ed switch).

jabranham commented:
This works well for me. Thanks, @akater! Looking forward to seeing your implementation of colored credits/debits!

cpitclaudel commented:
The width adjustment part looks nice; thanks! I have an alternate fix for appending parts of the command name.

I don't think we need `window-text-width`; window-width is just fine, but we do indeed need to subtract 1 to account for the continuation glyph column:

```
(window-width &optional WINDOW PIXELWISE)

Return the width of WINDOW’s text area.
WINDOW must be a live window and defaults to the selected one.  Optional
argument PIXELWISE non-nil means return the width in pixels.  The return
value does not include any vertical dividers, fringes or marginal areas,
or scroll bars.

If PIXELWISE is nil, return the largest integer smaller than WINDOW’s
pixel width divided by the character width of WINDOW’s frame.  This
means that if a column at the right of the text area is only partially
visible, that column is not counted.

Note that the returned value includes the column reserved for the
continuation glyph.
```

akater commented:
I am closing this pull request and recommend https://github.com/cpitclaudel/ledger-mode/tree/report-improvements as a replacement which covers ≈everything suggested here (and more), in a more robust way. The only thing that bothers me is the tty case. There might be something weird about my Emacs setup which necessitated using `(display-graphic-p)` to calculate window width.
